### PR TITLE
New conversation and getUserByUid

### PR DESCRIPTION
**Pull request by panivita**
# Description
Sorry for big PR. Here 2 task: Backend: get users should support getting user based on uid from Firebase #147 and Creating a new conversation #38. When I start work with #38  I discovered that in UserContext vi have user from firebase, vi don't have user.id from database. I decided don't changes my branch and continue work with 2 task in one branch. 
Task #147 in Users-controller.js I add getUserByUid. 
Task #38 in UserListComponent/UserList.js I add onclick function createConversation that:
1. find userId
2. check and redirect to exist channel
3. if users don't have channel create and redirect to new channel

Fixes # (issue)
I fixed api Getting channels with specific members. 
# How to test?
Task #147 http://localhost:3000/api/documentation/#/default/get_users_current
Task #38 http://localhost:3000/overview after log ind.
clicks on another user to create a new chat room or redirect to exist room
# Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] This PR is ready to be merged and not breaking any other functionality

### CODE REVIEW

File: src/client/components/UsersListComponent/UsersListStyle.css [hunk 1]
```diff
@@ -1,33 +1,20 @@
...
-  width: 70%;
-  height: 40%;
+.user-list img {
+  width: 20%;
   border-radius: 50%;
-}
-.user-list li span {
-  padding: 3px;
-}
+}
```
Comment (tpetrina) [inline src/client/components/UsersListComponent/UsersListStyle.css:20]: Missing new line.

File: src/server/api/controllers/channels-controller.js [hunk 2]
```diff
@@ -7,11 +7,7 @@ const createChannel = async (body) => {
   //  will return the total number of channels after inserting a new channel
   const insertedChannels = await knex('channels').insert(newChannel);
 
-  return {
-    successful: true,
-    //  id for the newly created channel will be same as total number of channels inserted
-    id: insertedChannels[0],
-  };
+  return { channelId: insertedChannels[0] };
```
Comment (tpetrina) [inline src/server/api/controllers/channels-controller.js:10]: It is a convention to return `id` here rather than `channelId`. You can always rename in the frontend.
I'd suggest you don't change this on the server side.

File: src/server/api/routes/channels-router.js [hunk 3]
```diff
@@ -7,7 +7,7 @@ const channelsController = require('../controllers/channels-controller');
 
 /**
  * @swagger
- * /modules:
+ * /channels:
```
Comment (tpetrina) [inline src/server/api/routes/channels-router.js:10]: Good catch!

File: src/server/api/controllers/channel-members.controller.js [hunk 4]
```diff
@@ -40,16 +42,23 @@ const createChannelMember = async (body) => {
...
-async function checkForCommoneChannels(arrOfUsers) {
-  // get all channels where the users are member off and the number of users is equal to arrOfUsers.length
-  const allChannelsForUsers = await knex('channel_members as a')
-    .count('a.fk_user_id', { as: 'numberOfUsers' })
-    .join('channel_members as b', 'b.fk_channel_id', 'a.fk_channel_id')
-    .select('a.id', 'a.fk_channel_id as channelId')
-    .whereIn('a.fk_user_id', arrOfUsers)
-    .groupBy('a.id')
-    .having('numberOfUsers', '=', arrOfUsers.length);
+const checkForCommoneChannels = async (arrOfUsers) => {
```
Comment (tpetrina) [inline src/server/api/controllers/channel-members.controller.js:45]: I'd rather you didn't fix this code and instead approached Wael to ensure his code works. Then your entire PR would be smaller and more concise.
Even though your code is much shorter.

**Review by tpetrina — CHANGES_REQUESTED** [review]
Overall good work, small things to fix here and there.

File: src/client/components/UsersListComponent/UsersList.js [hunk 5]
```diff
@@ -1,21 +1,71 @@
...
+    'Content-type': 'application/json',
+  },
+  body: JSON.stringify(body),
+});
+
 export default function Userslist({ users }) {
+  const history = useHistory();
+  const user = useContext(UserContext);
+
+  const createConversation = async (userId, currentUser) => {
```
[the remark this answers is not in the document]
Comment (panivita) [reply src/client/components/UsersListComponent/UsersList.js:24]: In container  should I create StartChat container or NewConversation and in component folder create StartChat component folder? Function createConversation I should moving to container?

File: src/server/api/controllers/channel-members.controller.js [hunk 6]
```diff
@@ -42,6 +42,16 @@ const createChannelMember = async (body) => {
...
+
+const checkForGeneralChannels = async (arrOfUsers) => {
+  const generalChannels = await knex('channel_members')
+    .select('fk_channel_id as channelId')
+    .whereIn('fk_user_id', arrOfUsers)
+    .groupBy('fk_channel_id')
+    .having(knex.raw('count(*)'), '=', arrOfUsers.length);
+  return generalChannels.map((channel) => channel.channelId);
+};
+
```
Comment (benna100) [inline src/server/api/controllers/channel-members.controller.js:54]: This is what @wael85 did in a seperate PR. This is why communication is **so** important

Comment (panivita) [reply src/server/api/controllers/channel-members.controller.js:54]: I know about that :) I tolk with Wael about that many times. But our api it is not the same.

File: src/client/components/Profile/Profile.stories.js [hunk 7]
```diff
@@ -1,10 +1,6 @@
 import React from 'react';
 import ProfileDatails from './ProfileDetails';
 
-export default {
```
Comment (tpetrina) [inline src/client/components/Profile/Profile.stories.js:4]: Why did you remove this one here? This is for Storybook.

File: src/client/components/Profile/Profile.stories.js [hunk 8]
```diff
@@ -13,8 +13,8 @@ const profileImage =
 export const profileView = () => (
   <ProfileDatails
     title="Test title"
```
[the remark this answers is not in the document]
Comment (tpetrina) [reply src/client/components/Profile/Profile.stories.js:11]: Title is still here!

File: src/client/api/getExistchannel.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+export async function getExistchannel(currentUser, userId) {
```
Comment (tpetrina) [inline src/client/api/getExistchannel.js:1]: Maybe a better name: `getExistingChannelForUsers`

File: src/client/components/UsersListComponent/UsersListStyle.css [hunk 10]
```diff
@@ -1,33 +1,20 @@
...
+  display: flex;
+  flex-direction: column;
   align-items: center;
-  text-align: center;
   width: 80%;
+  margin: 2% auto;
 }
-.user-list {
+
+.user-item {
```
Comment (tpetrina) [inline src/client/components/UsersListComponent/UsersListStyle.css:9]: You should scope this with `.user-list .user-item`. It's a good practice to do this always.

File: src/client/containers/UserList/onStartChat.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
+import { getExistchannel } from '../../api/getExistchannel';
+
+const requestConfig = (body) => ({
```
Comment (tpetrina) [inline src/client/containers/UserList/onStartChat.js:3]: You can actually remove this config and switch to `fetchWithAuth` instead. If you need help, let me know.

File: src/client/components/UsersListComponent/UsersList.js [hunk 12]
```diff
@@ -1,21 +1,36 @@
...
+import { useHistory } from 'react-router-dom';
 import PropTypes from 'prop-types';
+import { UserContext } from '../../context/userContext';
 import './UsersListStyle.css';
+import { onStartChat } from '../../containers/UserList/onStartChat';
 
 export default function Userslist({ users }) {
+  const history = useHistory();
+  const user = useContext(UserContext);
+  const createConversation = onStartChat(history);
```
Comment (tpetrina) [inline src/client/components/UsersListComponent/UsersList.js:15]: All of this should go to a `container component` and in there you use history, context, and handlers.
Components should preferably only contain HTML/CSS and use props to communicate with the parent container.
And components should never import anything from containers, only the other way around.

File: src/client/components/UsersListComponent/UsersListStyle.css (hunk 10 above)
Comment (panivita) [reply src/client/components/UsersListComponent/UsersListStyle.css:9]: I try read more why it is good practice use parent class and child class, but dont find. I think only because two class is more important than one class, if we have difference styling for the same element, styling with 2 css class will win. Mind you this, when you write about good practice? or why? https://css-tricks.com/a-use-case-for-a-parent-selector/

File: src/client/components/Profile/Profile.stories.js (hunk 7 above)
Comment (panivita) [reply src/client/components/Profile/Profile.stories.js:4]: carelessness

File: src/client/components/UsersListComponent/UsersList.js (hunk 12 above)
Comment (panivita) [reply src/client/components/UsersListComponent/UsersList.js:15]: I tried this, but I had problems 'React Hook  is called in function  which is neither a React function component or a custom React Hook function'. but now I resolved it. Mistake was here 'when you are using React Hooks, your functional component's name MUST start with a capital letter'. only should change one small ting - capital letter :(

File: src/client/containers/UserList/onStartChat.js (hunk 11 above)
Comment (panivita) [reply src/client/containers/UserList/onStartChat.js:3]: great idea :)

**Review by benna100 — APPROVED** [review]

**Comment by benna100** [comment]
Super strong code. The PR is a bit messy (which you also pointed out yourself) but still incredibly good job. This was not an easy task!
